### ISSUE TITLE
Merge Entities from multiple Org providers

### ISSUE THREAD

**Issue by rmanny** [backstage/backstage#11406]
The existing OrgProviders are perfect if all of the desired User and Group info exists in only one provider. But I think there are use cases where being able to use two in conjunction could fill some important holes

## Feature Suggestion

Add a strategy which would allow users to leverage information from two (or maybe more) OrgProviders when generating their hierarchy

## Possible Implementation

I think it will be necessary to schedule the two provider runs at the same time, lest we end up with a provider which needs to pull from existing entities in the catalog, so I am imagining an interface something like this when building the catalog processor:

```
  builder.addEntityProvider(
    MergedOrgEntityProvider.fromConfig(
      LdapOrgEntityProvider.fromConfig(env.config, {
        ...
        schedule: "inherit",
        ...
      }),
      GitHubOrgEntityProvider.fromConfig(env.config, {
        ...
        schedule: "inherit",
        ...
      }),
      {
        mergeOn: 'email',
        mergeStrategy: 'failOnCollision',
        schedule: env.scheduler.createScheduledTaskRunner({
          ...
        }),
      }
    )
  );
```

Then I think the merging could be hooked in right before the `this.connection.applyMutation` logic to kick the results up to the merged provider in a Promise.all to perform that step instead

A mergeStrategy of 'failOnCollision' is the only one that really makes sense to me, since I think it should be up to the user to set up the right logic inside their user and groupTransformers to select which properties they want

## Context

At my company, the org hierarchy is defined in Github. So to properly set up the membership relationships, that means we need to also import all of our Users from Github

However, all of our user metadata (full name, title, security groups, manager, etc) is stored in LDAP, and has no parallel inside Github

So, we are essentially torn between choosing the complete hierarchy info or the complete user info. Or of course creating a custom `FrankensteinsMonsterOrgProvider` which does both. Seems fragile and a lot of work for something that Backstage is already SO close to being able to do

**Comment by rmanny** [comment]
I have been pouring over the code for a while trying to figure out another way to do this, so I would be willing to take this task on if it seems like a good idea

**Comment by freben** [comment]
Hm, I do not think that a wrapper provider is the right way to go about it.

At a conceptual level, we commonly use providers to get "raw" data into the system and then use processors to enrich that data. Just looking at the context section of this ticket, it sounds like you'd just import whatever you have out of GitHub, and then you'd have a separate custom `EnrichUsersWithLdapDataProcessor` that reads (and caches) data out of LDAP and slaps it into those partial entities before they get persisted.

I'll also note that relations between users and groups are two-way. You can generate those edges in one of two ways: either the groups declare who their members are, or the users declare what groups they are members of. The end result is exactly the same. So it's perfectly fine to grab ONLY groups out of GitHub for example, if your users in LDAP know what group names they are members of (the mapping is "loose" and resolved internally by the catalog when stitching things together). Likewise, if you have all of the group AND membership info in GitHub, it's fine to import only the groups that way and leave the user references in the spec "dangling", and then import only users out of LDAP completely separately, unbeknownst to each other. Does that make sense?

Now, I know that the github import at this point has no option to just get groups or just get users. But that's easy to add, if this avenue is interesting to you.

**Comment by danopia** [comment]
Hi, thread :)

> you'd have a separate custom `EnrichUsersWithLdapDataProcessor` that reads (and caches) data out of LDAP and slaps it into those partial entities before they get persisted.

I tried going this route of preprocessing the User entities from the Github Org provider. Unfortunately, my org's github users are very unpredictably named, as they are personal usernames that everyone selected arbitrarily. I tried having my preprocessor rename the entities to be consistent and based on IT data and this failed because `The entity kind, namespace, or name changed during processing`.

I also looked at having Github only provide groups, but the Github Org provider only provides membership on the User entities. So if the Github user entities are skipped then the membership info is no longer available to the preprocessor and we end up with _just_ the team tree. I want to leverage the Github team membership as well!

In summary I see that preprocessors can add to an entity, but this isn't quite enough to make the github provider useful as is. I'd love some way to either rename entities or give the Github Org provider a mapping of username -> desired name. The presence of a [`github.com/user-login`](https://backstage.io/docs/features/software-catalog/well-known-annotations#githubcomuser-login) annotation feels like it expects username mapping to occur.

**Comment by rmanny** [comment]
@danopia I have been dealing with the same mapping question. If the provider would pull the email, that would solve most of the immediate problems (catalog user resolution, AD Groups, etc). I haven't looked into why this doesn't happen yet, I assume it's a limitation of the API, email is relatively sensitive

**Comment by freben** [comment]
> `The entity kind, namespace, or name changed during processing`.

Yeah this is an unavoidable limitation at this point, since it's part of the actual primary ID / key of the entity. Everything hangs off of that kind/namespace/name triplet so it can't change in-flight once it's set (the very database schema is based on it) because the end result would be an entirely other entity.

> I also looked at having Github only provide groups, but the Github Org provider only provides membership on the User entities. So if the Github user entities are skipped then the membership info is no longer available to the preprocessor and we end up with _just_ the team tree. I want to leverage the Github team membership as well!

Oh that's unfortunate that they don't go the other way from the group end :/ Not sure what to do about that one besides getting it all in an entirely custom provider and modifying the dataset in memory before emitting it all. Then again ... looking at [the graphql code](https://github.com/backstage/backstage/blob/179a5906560a8d5b2cc4ff3b229dd76a91387eca/plugins/catalog-backend-module-github/src/lib/github.ts), I wonder if it wouldn't be possible to retrieve both ways?

> In summary I see that preprocessors can add to an entity, but this isn't quite enough to make the github provider useful as is. I'd love some way to either rename entities or give the Github Org provider a mapping of username -> desired name. The presence of a [`github.com/user-login`](https://backstage.io/docs/features/software-catalog/well-known-annotations#githubcomuser-login) annotation feels like it expects username mapping to occur.

I see that the [GitHubOrgEntityProvider](https://github.com/backstage/backstage/blob/179a5906560a8d5b2cc4ff3b229dd76a91387eca/plugins/catalog-backend-module-github/src/GitHubOrgEntityProvider.ts#L52) doesn't even have the ability to transform things as you see fit before they are emitted like [in the ldap one](https://github.com/backstage/backstage/blob/179a5906560a8d5b2cc4ff3b229dd76a91387eca/plugins/catalog-backend-module-ldap/src/processors/LdapOrgEntityProvider.ts#L86). But I guess that wouldn't be sufficient in this case either way 🤔 

But OK, backing up a bit. @rmanny it sounds to me like you are stuck in a place where you may want to do the whole translation yourself, upfront, in a custom entity provider. The names being different, AND them coming from two different systems, feels hard to overcome in any standard way at the moment.

Maybe if you do start on this trek, there might be something that could be exported out of e.g. `@backstage/plugin-catalog-backend-module-github` to make it more of a oneliner to leverage the existing reader code? Instead of rewriting it all.

**Comment by freben** [comment]
Hm, I'm not entirely sure I fully understand what you mean with merging and deleting, or how you meant with "when the user logs in". 🤔 

Sounds to me like you would be best served by writing an entirely custom entity provider from scratch that does exactly what you need. Hopefully the existing catalog modules export enough lower level building blocks to help you get started.

Unless your case is straight forward in such a way that you can cleanly import all groups from one place and all users from another place and they happen to line up in their naming etc.

**Comment by angeliski** [comment]
> Hm, I'm not entirely sure I fully understand what you mean with merging and deleting

In my use case I would just ingest Github and Okta in my catalog, so when the user try to [autheticate](https://backstage.io/docs/auth/#configuring-authentication-providers) in backstage using Github or Okta there will be there.

 When I say `merging and deleting` I just want to get this to "different entities" and make one merged.
Like `UserAngeliskiGithub` and `UserAngeliskiOkta` transform in `UserAngeliskiGithub` with okta information.

> or how you meant with "when the user logs in"

Here I just want to use someting like [AuthHandler](https://backstage.io/docs/auth/identity-resolver#authhandler) to connect those two entities in the catalog.

>Sounds to me like you would be best served by writing an entirely custom entity provider from scratch that does exactly what you need. Hopefully the existing catalog modules export enough lower level building blocks to help you get started.

I am trying to avoid that 😆 Because those two entites are very differentes (we use personal account in Github) and it's hard to connect. 
My integration point will be the user identity (when they use Github and Okta, I connect those entities). This could work for other points to, like Slack users.

Maybe I can change the question:
How can I connect two User entities in my catalog to say: "Hey, those two users are one"
Like we can do in Velocity (Ref: https://docs.velocity.codeclimate.com/en/articles/1708709-de-duplicating-contributors)

**Comment by freben** [comment]
> Because those two entites are very differentes (we use personal account in Github) and it's hard to connect.

I'll start by recognizing this sentence, and this is a fact no matter what. They are different worlds and there's no existing edge between them.

> My integration point will be the user identity (when they use Github and Okta, I connect those entities)

So at that point, after performing the oauth dance with github, you will have more knowledge - by the fact that you know who the okta user is that performed the dance. And you will want to persist that fact so that you can remember it for the future.

It's not necessarily the entities themselves that need to be changed. Your ultimate goal, I presume, is to be able to add the github user to the user's identity information in the Backstage token on sign-in.

But you do want to persist the "edge" of okta user -> github user. And I think you will have to do so in an external system. Maybe in a tiny Backstage backend plugin that has a single mapping table. Your okta (main sign-in) signIn.resolver will read info from that plugin, and signing in persists info in that plugin.

Does that make sense?

**Comment by angeliski** [comment]
> It's not necessarily the entities themselves that need to be changed. Your ultimate goal, I presume, is to be able to add the github user to the user's identity information in the Backstage token on sign-in.

This is very interesting. So you are saying I don't need **merge entities** just put the info in the Backstage token 🤔 

Maybe this could create a chaotic catalog, because in some cases my service is owned by "user a-github" and others by "user a-okta"

> But you do want to persist the "edge" of okta user -> github user. And I think you will have to do so in an external system. Maybe in a tiny Backstage backend plugin that has a single mapping table. Your okta (main sign-in) signIn.resolver will read info from that plugin, and signing in persists info in that plugin.
>
> Does that make sense?

Yes, I **really** want to merge those two entities, but I got your perspective now, totally makes sense. 
When we start to solve this problem I will send a PR to discuss more about that, thanks for the advice @freben

**Comment by freben** [comment]
Alright!

And I'd add that if you WERE to actually want to produce merged user entities inside the catalog (maybe some profile data from place A, and the name from place B etc), then you really would have to do that merging "outside" anyway. Either in a custom entity provider that queries multiple systems, or in a completely external standalone service which you ingest from.

It may be interesting to know that we use the latter approach. We have a team that owns the org data aspects, not for Backstage but for the organization as a whole. And they have services to query, that hide all of the complexity behind the scenes.

**Comment by benkeil** [comment]
We wanted to enrich the group data with data our company doesn't save at all at the moment.
We decided to create a new Entity `GroupInfo` and put it in the `catalog-info.yaml`.
Then we added a custom `CatalogProcessor` that adds relations between the entities. You can then query relations of the group to get these data.
But not sure if this is a good approach...

**Comment by robbat2** [comment]
For future readers here: https://roadie.io/docs/catalog/decorating-components/ provides a way forward (I am not a Roadie Customer, but it looks promising as a way to solve the issue asked)
